wasm-rpc/src/lib.rs

```diff
@@ -304,6 +402,576 @@ fn build_tree(node: &WitNode, nodes: &[WitNode]) -> Value {
     }
 }
 
+// An efficient read representation of a value with type information at every node
+// Note that we do an extra annotation of AnalysedType at complex structures to fetch their type information with 0(1)
+// However the typical use of TypeAnnotatedValue is similar to using a Json/Yaml etc
+#[derive(Clone, Debug)]
+pub enum TypeAnnotatedValue {
+    Bool(bool),
+    S8(i8),
+    U8(u8),
+    S16(i16),
+    U16(u16),
+    S32(i32),
+    U32(u32),
+    S64(i64),
+    U64(u64),
+    F32(f32),
+    F64(f64),
+    Chr(char),
+    Str(String),
+    List(ListValue),
+    Tuple(TupleValue),
+    Record(RecordValue),
+    Flags(FlagValue),
+    Variant(VariantValue),
+    Enum(EnumValue),
+    Option(OptionValue),
+    Result(ResultValue),
+    Handle(ResourceValue),
+}
+
+#[derive(Clone, Debug)]
+pub struct EnumValue {
```

Comment:
Why do we need these separate structures instead of just having fields in the enum cases of `TypedAnnotatedValue`? 

And if there is a good reason, can we make them non public?

Reply:
At every node it also keeps track of the total analysed_type as well not just the type of the value. Example: What is are the all possibilites of `enum` along with holding the actual `String`. `TypeAnnotatedValue` can be usually formed mostly through `AnalysedType` and `Value` and there is no need to forget this information. Also it keeps a restriction to what can be formed with TypeAnnotatedValue. Example: `List(ListValue)` , where `ListValue` holds the type of the list and also `Vec<TypeAnnotatedValue>` instead of just `List(Vec<TypeAnnotatedValue)`.  The latter sort of indicates list can hold different types.

Let me see if i can make it private, but i feel it could be useful, in printing errors on expected types etc.

Reply:
I am happy to remove this, but I feel we would come back and add it again.

Reply:
I'm only asking why not:
```rust
enum TypedAnnotatedValue {
  // ...
   Enum {
     // fields of what currently is EnumValue
   }
}
```

Reply:
Done